pkg/proxy/iptables/proxier.go

```diff
@@ -639,6 +640,17 @@ func (proxier *Proxier) OnNodeUpdate(oldNode, node *v1.Node) {
 		return
 	}
 
+	// kube-proxy in LocalModeNodeCIDR mode may cache stale Node.PodCIDR.
+	// It should restart on those situations to reconcile the state.
+	// Ref: https://issues.k8s.io/111321
+	if proxier.localDetector.IsImplemented() && proxier.localDetector.Mode() == proxyconfigapi.LocalModeNodeCIDR {
+		nodeLocalCIDR := proxier.localDetector.IfLocal()[1]
+		if !utilproxy.NodeContainsPodCIDR(node, nodeLocalCIDR) {
+			klog.ErrorS(nil, "Current Node PodCIDR(s) not longer matches the PodCIDR kube-proxy saw at startup, restarting", "node", klog.KObj(node), "local CIDR", nodeLocalCIDR, "PodCIDRs", node.Spec.PodCIDRs)
```

Comment:
This restarts if _any_ node's PodCIDR doesn't match, doesn't it? You need to check `node.Name` or something...

But also, rather than checking the PodCIDR specifically, kube-proxy could just restart if its own Node gets deleted, or if its `UID` or `creationTimestamp` changes. (I think the shared informer code will potentially squash a delete+create into an update like that?)

Reply:
this happens at startup
 
https://github.com/kubernetes/kubernetes/blob/70dcb0f1297e0a7b183720cfacece73c57ee1afc/cmd/kube-proxy/app/server_others.go#L166-L173

Reply:
hold on, your into something, that is not the informer code, that is the code that checks for the PodCIDR on the nodeName before start.

The informer runs only on the node name

https://github.com/kubernetes/kubernetes/blob/d96af5f2761c23eddde206aee7b7147a9692b737/cmd/kube-proxy/app/server.go#L753-L766


.... but only works with TopologyAwareHints feature gate enabled @bowei